examples/ex_color2.c

```diff
@@ -0,0 +1,348 @@
+#define ALLEGRO_UNSTABLE
+#include <allegro5/allegro.h>
+#include <allegro5/allegro_font.h>
+#include <allegro5/allegro_color.h>
+#include <allegro5/allegro_primitives.h>
+#include <stdlib.h>
+#include <math.h>
+
+#include "common.c"
+
+#define FPS 60
+
+typedef struct {
+   ALLEGRO_COLOR rgb;
+   float l, a, b;
+} Color;
+
+struct Example {
+   ALLEGRO_FONT *font;
+   ALLEGRO_BITMAP *lab[512];
+   ALLEGRO_COLOR black;
+   ALLEGRO_COLOR white;
+   Color color[2];
+   int mx, my;
+   int mb; // 1 = clicked, 2 = released, 3 = pressed
+   int slider;
+   int top_y;
+   int left_x;
+   int slider_x;
+   int half_x;
+} example;
+
+static void draw_lab(int l)
+{
+   if (example.lab[l]) return;
+
+   example.lab[l] = al_create_bitmap(512, 512);
+   ALLEGRO_LOCKED_REGION *rg = al_lock_bitmap(example.lab[l],
+         ALLEGRO_PIXEL_FORMAT_ABGR_8888_LE, ALLEGRO_LOCK_WRITEONLY);
+   int x, y;
+   for (y = 0; y < 512; y++) {
+      for (x = 0; x < 512; x++) {
+         float a = (x - 511.0 / 2) / (511.0 / 2);
+         float b = (y - 511.0 / 2) / (511.0 / 2);
+         float rf, gf, bf;
+         b = -b;
+         al_color_lab_to_rgb(l / 511.0, a, b, &rf, &gf, &bf);
+         if (rf < 0 || rf > 1 || gf < 0 || gf > 1 || bf < 0 || bf > 1) {
+            rf = gf = bf = 0;
+         }
+         int red = 255 * rf;
+         int green = 255 * gf;
+         int blue = 255 * bf;
+
+         uint8_t *p = rg->data;
+         p += rg->pitch * y;
+         p += x * 4;
+         p[0] = red;
+         p[1] = green;
+         p[2] = blue;
+         p[3] = 255;
+      }
+   }
+   al_unlock_bitmap(example.lab[l]);
+}
+
+static void draw_range(int ci)
+{
+   float l = example.color[ci].l;
+   int cx = (example.color[ci].a * 511.0 / 2) + 511.0 / 2;
+   int cy = (-example.color[ci].b * 511.0 / 2) + 511.0 / 2;
+   int r = 2;
+   /*
+    * 1234561
```

Comment:
I need to comment this. Basically we loop around the selected color in the center to find the pixels with a certain ciede2000 distance. We start around the center then go 1, 2 to the right. Then we go 1, 2 down. Then 1, 2 left. Then 1, 2 up. Then we jump to a bigger radius and start over with 1, 2, 3, 4 right, 1, 2, 3, 4 down... The diagram shows how exactly :)